Sources/NIOOpenSSL/SSLPKCS12Bundle.swift

```diff
@@ -0,0 +1,218 @@
+///
+///    let p12Bundle = OpenSSLPKCS12Bundle(file: pathToMyP12)
+///    let config = TLSConfiguration.forServer(certificateChain: p12Bundle.certificateChain,
+///                                            privateKey: p12Bundle.privateKey)
+///
+/// The created `TLSConfiguration` can then be safely used for your endpoint.
+public struct OpenSSLPKCS12Bundle {
+    public let certificateChain: [OpenSSLCertificate]
+    public let privateKey: OpenSSLPrivateKey
+
+    private init<Bytes: Collection>(ref: UnsafeMutablePointer<PKCS12>, passphrase: Bytes?) throws where Bytes.Element == UInt8 {
+        var pkey: UnsafeMutablePointer<EVP_PKEY>? = nil
+        var cert: UnsafeMutablePointer<X509>? = nil
+        var caCerts: UnsafeMutablePointer<stack_st_X509>? = nil
+
+        let rc = try passphrase.withSecureBytes { passphrase in
+            PKCS12_parse(ref, passphrase, &pkey, &cert, &caCerts)
+        }
+        guard rc == 1 else {
+            throw OpenSSLError.unknownError(.init())
+        }
+
+        // Successfully parsed, let's unpack. The key and cert are mandatory,
+        // the ca stack is not.
+        guard let actualCert = cert, let actualKey = pkey else {
+            // Free the pointers that we have.
+            X509_free(cert)
+            EVP_PKEY_free(pkey)
+            CNIOOpenSSL_sk_X509_free(caCerts)
+            throw NIOOpenSSLError.unableToAllocateOpenSSLObject
+        }
+
+        let certStackSize = caCerts.map(CNIOOpenSSL_sk_X509_num) ?? 0
+        var certs = [OpenSSLCertificate]()
+        certs.reserveCapacity(Int(certStackSize) + 1)
+        certs.append(OpenSSLCertificate.fromUnsafePointer(pointer: actualCert))
+
+        for idx in 0..<certStackSize {
+            guard let stackCertPtr = CNIOOpenSSL_sk_X509_value(caCerts, idx) else {
+                preconditionFailure("Unable to get cert \(idx) from stack \(String(describing: caCerts))")
+            }
+            certs.append(OpenSSLCertificate.fromUnsafePointer(pointer: stackCertPtr))
+        }
+
+        self.certificateChain = certs
+        self.privateKey = OpenSSLPrivateKey.fromUnsafePointer(pointer: actualKey)
+
+    }
+
+    /// Create a `OpenSSLPKCS12Bundle` from the given bytes in memory,
+    /// optionally decrypting the bundle with the given passphrase.
+    ///
+    /// - parameters:
+    ///     - buffer: The bytes of the PKCS#12 bundle.
+    ///     - passphrase: The passphrase used for the bundle, as a sequence of UTF-8 bytes.
+    public init<Bytes: Collection>(buffer: [UInt8], passphrase: Bytes?) throws where Bytes.Element == UInt8 {
+        guard openSSLIsInitialized else { fatalError("Failed to initialize OpenSSL") }
+        
+        let p12 = buffer.withUnsafeBytes { pointer -> UnsafeMutablePointer<PKCS12>? in
+            let bio = BIO_new_mem_buf(UnsafeMutableRawPointer(mutating: pointer.baseAddress), CInt(pointer.count))!
+            defer {
+                BIO_free(bio)
```

Comment:
so `bio` can be freed before the returned value?

Reply:
Yes. The BIO is used to read the data in, but that's it: the PKCS12 object holds a structured version of the data.